### PR TITLE
✨ add Settings support to RuntimeExtensions

### LINKED ISSUE #7432: Support settings for RuntimeExtensions
**User Story**

As a RuntimeExtensions author, I would like to offer settings allowing cluster administrators to tune my RuntimeExtensions behavior when installed on a management cluster

As a RuntimeExtensions author, I would like to offer settings allowing cluster class authors to tune my RuntimeExtensions behavior when used in a ClusterClass

**Detailed Description**

RuntimeExtension is a powerful mechanism for extending Cluster API and we expect RuntimeExtensions to become more and more complex over time. This will lead RuntimeExtensions authors to expose settings allowing users to customize RuntimeExtensions behavior on different clusters or when used in different ClusterClass.

Eg. I have Runtime extension with a couple of lifecycle hooks, and on specific management clusters I want to disable some of them or simply tune the RequeAfter interval

Eg. I have Runtime extension with many external patches, and on specific ClusterClasses I want to disable some of them

RuntimeExtension extension can implement those use cases in many ways, exposing flags, checking for setting in a config map etc. etc. but all those approaches won't be integrated with Cluster API (so the users will be required to change RuntimeExtension deployment or to act on different objects).

With this issue we are proposing to introduce a setting fields both in 

```yaml
apiVersion: runtime.cluster.x-k8s.io/v1alpha1
kind: ExtensionConfig
metadata:
  name: test-extension
spec:
  # settings allows to define key value pairs
  # to be passed to the all the calls to RuntimeExtensions provided
  # by the this component.
  settings:
      disableBeforeCreateClusterHook: "true"
      retryAfterInterval: "20s"
```

And in 

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
metadata:
  name: quick-start-runtimesdk
spec:
  ...
  variables:
      <<- for the user / they surface on the Cluster
  
  patches:
  - name: lbImageRepository
    external:
      # settings allows to define key value pairs to be passed to the topology mutation hook call;
      # settings provided in this section will be merged with settings from the ExtensionConfig, with settings
      # defined here taking precedence in case of conflicts.
      # NOTE: settins are visible to the ClusterClass authors only and cannot be changed on a Cluster bases
      # like variables.
      settings:
          disableHttpProxyPatch: "true"
      generateExtension: generate-patches.k8s-upgrade-with-runtimesdk
      validateExtension: validate-topology.k8s-upgrade-with-runtimesdk
```

/kind feature
/area runtime-sdk

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR adds `Settings` support to RuntimeExtensions. Settings can be used to send arbitrary key value pairs in the request body of the call to the runtime extension.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7432

### CODE REVIEW

File: exp/runtime/hooks/api/v1alpha1/common_types.go [hunk 1]
```diff
@@ -20,6 +20,31 @@ import (
...
+type RequestObject interface {
+	runtime.Object
+	GetSettings() map[string]string
+	SetSettings(settings map[string]string)
+}
+
+// CommonRequest is the data structure common to all request types.
+// Note: By embedding CommonRequest in a runtime.Object the RequestObject
+// interface is satisfied.
+type CommonRequest struct {
```
Comment (fabriziopandini) [inline exp/runtime/hooks/api/v1alpha1/common_types.go:34]: just noticing now that we have a small inconsistency between `CommonRequest`, `CommonResponse`, `ResponseStatus`, `CommonRetryResponse`, and `RetryResponseObject` (which has an Object suffix and doesn't have a Common prefix). Wondering if we should/could fix this in a follow-up PR
@killianmuldoon @sbueringer opinions?

File: exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go [hunk 2]
```diff
@@ -28,6 +28,9 @@ import (
 type BeforeClusterCreateRequest struct {
 	metav1.TypeMeta `json:",inline"`
 
+	// CommonRequest contains Settings field common to all request types.
```
Comment (fabriziopandini) [inline exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go:31]: What about keeping this generic so it won't become outdated if we add more fields to CommonRequest, e.g
```suggestion
	// CommonRequest contains fields common to all request types.
```

File: internal/controllers/topology/cluster/patches/external/external_patch_generator.go [hunk 3]
```diff
@@ -48,6 +48,10 @@ func (e externalPatchGenerator) Generate(ctx context.Context, forObject client.O
 	if !feature.Gates.Enabled(feature.RuntimeSDK) {
 		return nil, errors.Errorf("can not use external patch %q if RuntimeSDK feature flag is disabled", *e.patch.External.GenerateExtension)
 	}
+
+	// Set the settings defined in external patch definition on the request object.
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/patches/external/external_patch_generator.go:52]: Let's add a note that those settings applies on top of settings defined on the ExtensionConfig

File: internal/runtime/client/client.go [hunk 4]
```diff
@@ -340,6 +340,20 @@ func (c *client) CallExtension(ctx context.Context, hook runtimecatalog.Hook, fo
...
+	settings := request.GetSettings()
+	if settings == nil {
+		settings = map[string]string{}
+	}
+	for k, v := range registration.Settings {
+		if _, ok := settings[k]; !ok {
+			settings[k] = v
+		}
+	}
+	request.SetSettings(settings)
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:355]: I think we should avoid side effects on the request object, what about deep copying it before updating settings? (also this should be covered by the test)

File: internal/runtime/client/client.go [hunk 5]
```diff
@@ -340,6 +340,20 @@ func (c *client) CallExtension(ctx context.Context, hook runtimecatalog.Hook, fo
...
+	// The values in request take precedence over the values in the registration.
+	settings := request.GetSettings()
+	if settings == nil {
+		settings = map[string]string{}
+	}
+	for k, v := range registration.Settings {
+		if _, ok := settings[k]; !ok {
+			settings[k] = v
+		}
+	}
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:354]: Nit in order to avoid the negation in the inner if
```suggestion
	settings := registration.Settings
	if settings == nil {
		settings = map[string]string{}
	}
	for k, v := range request.GetSettings() {
		settings[k] = v
	}
```

File: internal/controllers/topology/cluster/patches/external/external_patch_generator.go [hunk 6]
```diff
@@ -48,6 +48,10 @@ func (e externalPatchGenerator) Generate(ctx context.Context, forObject client.O
...
+	req.Settings = e.patch.External.Settings
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/patches/external/external_patch_generator.go:53]: What about adding a test for settings to exist in the generated call?

File: internal/runtime/client/client.go (hunk 5 above)
Comment (fabriziopandini) [reply internal/runtime/client/client.go:354]: What about adding a test for settings to exist/be properly merged in the generated call?

**Review by fabriziopandini — COMMENTED** [review]
PR looks good to me (mostly nits)
What about adding a stronger test signal by addressing https://github.com/kubernetes-sigs/cluster-api/issues/7322#issuecomment-1320392461 within this PR as well? (feel free to say no)
@sbueringer @killianmuldoon opinions on ^^

**Comment by sbueringer** [comment]
> What about adding a stronger test signal by addressing https://github.com/kubernetes-sigs/cluster-api/issues/7322#issuecomment-1320392461 within this PR as well? (feel free to say no)
> @sbueringer @killianmuldoon opinions on ^^

Sounds good to me

File: exp/runtime/hooks/api/v1alpha1/common_types.go (hunk 1 above)
Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/common_types.go:34]: This was intentional. Our interfaces are called ResponseObject and RetryResponseObject (infered roughly from runtime.Object which is embedded).

Then we have CommonResponse and CommonRetryResponse which are structs that are shared across multiple response structs

Comment (sbueringer) [reply exp/runtime/hooks/api/v1alpha1/common_types.go:34]: So I would keep it as is.

Changing it should also bring up the question of if we have to bump to v1alpha2

Comment (ykakarap) [reply exp/runtime/hooks/api/v1alpha1/common_types.go:34]: +1 to the current naming scheme.

**Comment by ykakarap** [comment]
>PR looks good to me (mostly nits)
What about adding a stronger test signal by addressing https://github.com/kubernetes-sigs/cluster-api/issues/7322#issuecomment-1320392461 within this PR as well? (feel free to say no)
@sbueringer @killianmuldoon opinions on ^^

Sounds good to me. I will make the change in this PR.

**Comment by sbueringer** [comment]
Sounds good.

Please let me know if/when I should review the PR, as usual I'm ignoring WIP PRs except if I'm asked for feedback/review.

**Comment by ykakarap** [comment]
@sbueringer This is ready for review.

File: exp/runtime/hooks/api/v1alpha1/common_types.go [hunk 7]
```diff
@@ -20,6 +20,31 @@ import (
...
+	Settings map[string]string `json:"settings"`
```
Comment (sbueringer) [inline exp/runtime/hooks/api/v1alpha1/common_types.go:35]: Should this have +optional + omitempty?

File: internal/controllers/topology/cluster/patches/external/external_patch_generator.go [hunk 8]
```diff
@@ -48,6 +48,11 @@ func (e externalPatchGenerator) Generate(ctx context.Context, forObject client.O
 	if !feature.Gates.Enabled(feature.RuntimeSDK) {
 		return nil, errors.Errorf("can not use external patch %q if RuntimeSDK feature flag is disabled", *e.patch.External.GenerateExtension)
 	}
+
+	// Set the settings defined in external patch definition on the request object.
+	// These settings will override overlapping keys defined in ExtensionConfig settings.
+	req.Settings = e.patch.External.Settings
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/external/external_patch_generator.go:54]: Given that we are re-using req for all patches. I think we should set req.Settings to nil after the call (maybe defer to capture all cases?)

File: internal/controllers/topology/cluster/patches/external/external_patch_generator_test.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+	runtimecatalog "sigs.k8s.io/cluster-api/exp/runtime/catalog"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/feature"
+	runtimeclient "sigs.k8s.io/cluster-api/internal/runtime/client"
+)
+
+var _ runtimeclient.Client = &fakeRuntimeClient{}
+
+type fakeRuntimeClient struct {
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/external/external_patch_generator_test.go:38]: nit: let's please move this test util down below the actual test

File: internal/controllers/topology/cluster/patches/external/external_patch_generator.go [hunk 10]
```diff
@@ -48,6 +48,11 @@ func (e externalPatchGenerator) Generate(ctx context.Context, forObject client.O
 	if !feature.Gates.Enabled(feature.RuntimeSDK) {
 		return nil, errors.Errorf("can not use external patch %q if RuntimeSDK feature flag is disabled", *e.patch.External.GenerateExtension)
 	}
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/external/external_patch_generator.go:50]: I think we need ~ the same in external_validator.go

File: internal/runtime/client/client.go [hunk 11]
```diff
@@ -340,6 +340,10 @@ func (c *client) CallExtension(ctx context.Context, hook runtimecatalog.Hook, fo
 	if registration.TimeoutSeconds != nil {
 		timeoutDuration = time.Duration(*registration.TimeoutSeconds) * time.Second
 	}
+
+	// Prepare the request by merging the settings in the registration with the settings in the request.
+	request = prepareRequest(request, registration)
```
Comment (sbueringer) [inline internal/runtime/client/client.go:345]: Let's please use a more specific func name

File: internal/runtime/client/client.go [hunk 12]
```diff
@@ -382,6 +386,23 @@ func (c *client) CallExtension(ctx context.Context, hook runtimecatalog.Hook, fo
...
 }
 
+// prepareRequest creates a new request object by adding the appropriate values from the registration.
+func prepareRequest(request runtimehooksv1.RequestObject, registration *runtimeregistry.ExtensionRegistration) runtimehooksv1.RequestObject {
+	// Merge the settings from registration with the settings in the request.
+	// The values in request take precedence over the values in the registration.
+	// Create a deepcopy object to avoid side-effects on the request object.
+	request = request.DeepCopyObject().(runtimehooksv1.RequestObject)
+	settings := registration.Settings
+	if settings == nil {
```
Comment (sbueringer) [inline internal/runtime/client/client.go:396]: Let's create a new map and then add key/value pairs from both settings.

The current code will modify the map inside of the registration which means that settings of all calls will add up in the registration

File: internal/runtime/client/client.go [hunk 13]
```diff
@@ -382,6 +386,23 @@ func (c *client) CallExtension(ctx context.Context, hook runtimecatalog.Hook, fo
...
+		settings = map[string]string{}
+	}
+	for k, v := range request.GetSettings() {
+		settings[k] = v
+	}
+	request.SetSettings(settings)
+	return request
```
Comment (sbueringer) [inline internal/runtime/client/client.go:403]: Maybe let's not also return the request. Modifying the input parameter + returning is always a bit confusing (as it's unclear if the input parameter is modified for the consumer). So I would only do it if it's necessary for the way a func is used

File: internal/runtime/client/client.go (hunk 12 above)
Comment (sbueringer) [reply internal/runtime/client/client.go:396]: Let's make sure we cover this edge case with a unit test
(we had this bug in various parts of our case base already)

File: test/extension/handlers/lifecycle/handlers.go [hunk 14]
```diff
@@ -176,13 +176,19 @@ func (m *ExtensionHandlers) DoBeforeClusterDelete(ctx context.Context, request *
...
 	hookName := runtimecatalog.HookName(hook)
 	configMap := &corev1.ConfigMap{}
 	configMapName := fmt.Sprintf("%s-test-extension-hookresponses", cluster.Name)
 	if err := m.client.Get(ctx, client.ObjectKey{Namespace: cluster.Namespace, Name: configMapName}, configMap); err != nil {
 		if apierrors.IsNotFound(err) {
-			configMap = responsesConfigMap(cluster)
+			// A ConfigMap of responses does not exist. Create one now.
+			// The ConfigMap is created with blocking responses if "defaultAllHandlersToBlocking" is set to "true"
+			// in the settings.
+			// This allows the test-extension to have non-blocking behavior by default byt can be switched to blocking
```
Comment (sbueringer) [inline test/extension/handlers/lifecycle/handlers.go:188]:
```suggestion
			// This allows the test-extension to have non-blocking behavior by default but can be switched to blocking
```

**Review by sbueringer — COMMENTED** [review]
Really nice work

A few findings, but nothing  big.

File: api/v1beta1/clusterclass_types.go [hunk 15]
```diff
@@ -532,6 +532,12 @@ type ExternalPatchDefinition struct {
 	// ValidateExtension references an extension which is called to validate the topology.
 	// +optional
 	ValidateExtension *string `json:"validateExtension,omitempty"`
+
+	// Settings defines key value pairs to be passed to mutation hook.
+	// Value defined here take precedence over the values defined in the
+	// corresponding ExtensionConfig.
+	// +optional
+	Settings map[string]string `json:"settings,omitempty"`
```
Comment (sbueringer) [inline api/v1beta1/clusterclass_types.go:540]: @ykakarap Can you please look into this?

> Incompatible changes:
>  - ExternalPatchDefinition: old is comparable, new is not
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/7562/pull-cluster-api-apidiff-main/1595553240982753280

Not sure if it is relevant

**Review by fabriziopandini — COMMENTED** [review]
lgtm for me pending comments

File: internal/controllers/topology/cluster/patches/external/external_patch_generator.go (hunk 8 above)
Comment (ykakarap) [reply internal/controllers/topology/cluster/patches/external/external_patch_generator.go:54]: Good catch! 
The next call to the generator should reset the settings to the value of that patch call but it is safer to reset it with in the same call.

File: internal/runtime/client/client.go (hunk 13 above)
Comment (ykakarap) [reply internal/runtime/client/client.go:403]: Here the input request is not modified (by design to avoid side effects when a user calls the CallExtension function). Therefore we have to return the request object here.

File: api/v1beta1/clusterclass_types.go (hunk 15 above)
Comment (ykakarap) [reply api/v1beta1/clusterclass_types.go:540]: This is expected. Maps are not comparable in go. Since a map (Settings) was added to `ExternalPatchDefinition` the entire struct now become un-comparable. This should be fine as we don't really ever do operations like `externalPatchDefinition1 == externalPatchDefinition2`.

**Comment by fabriziopandini** [comment]
great work!
/lgtm
/approved
